R/dictionary_based_cleaning.R

```diff
@@ -0,0 +1,228 @@
+#' Convert Redcap data dictionary into {matchmaker} dictionary format
+#'
+#' @param metadata a `data frame` with the data dictionary associated to a
+#'    Redcap project
+#' @param field_column the name of the column, in the input dictionary, that
+#'    contains the field names in of the Redcap project data
+#' @param opt_column the name of the column, in the input dictionary, that
+#'    contains the definition of the choices in every column of the Redcap
+#'    project data
+#' @param field_type a `character` with the name of the column that contains the
+#'    field type information
+#'
+#' @return a data frame with 4 columns. This is in the format required by the
+#'    {matchmaker} R package for dictionary-based cleaning.
+#' @keywords internal
+#' @noRd
+#'
+#' @examples
+#' test <- make_readcap_dictionary(
+#'   metadata     = system.file("extdata", "test_readcap_dictionary.RDS",
+#'                              package = "cleanepi"),
+#'   field_column = "field_name",
+#'   opt_column   = "select_choices_or_calculations",
+#'   field_type   = "field_type"
+#' )
+#'
+#'
+make_readcap_dictionary <- function(metadata,
+                                    field_column,
+                                    opt_column,
+                                    field_type) {
+  checkmate::assert_data_frame(metadata, min.rows = 1L, min.cols = 1L,
+                               null.ok = FALSE)
+  checkmate::assert_character(opt_column, len = 1L, null.ok = FALSE)
+  checkmate::assert_character(field_type, len = 1L, null.ok = FALSE)
+  checkmate::assert_character(field_column, len = 1L, null.ok = FALSE)
+
+  stopifnot(opt_column %in% names(metadata))
+
+  metadata     <- metadata %>%
+    dplyr::filter(!is.na(metadata[[opt_column]]),
+                  metadata[[field_type]] != "calc")
+
+  dictionary   <- NULL
+  for (i in seq_len(nrow(metadata))) {
+    dictionary <- rbind(dictionary,
+                        make_metadata(metadata[[opt_column]][i],
+                                      metadata[[field_column]][i]))
+  }
+  dictionary
+}
+
+#' Make data dictionary for 1 field
+#'
+#' @param x a `vector` choices from the input data
+#' @param field_column a `character` with the name of the field where the
+#'    choices are made.
+#'
+#' @return a `data frame` with the dictionary in the format that is accepted by
+#'    the {matchmaker} package.
+#' @keywords internal
+#' @noRd
+#'
+make_metadata <- function(x, field_column) {
+  splits          <- unlist(strsplit(x, "|", fixed = TRUE))
+  splits          <- trimws(splits)
+  combined_splits <- lapply(splits, get_meta_rows)
```

Comment:
Consider defining the function inline instead of separately.

```
combined_splits <- lapply(splits, function(x) {
  trimws(unlist(strsplit(x, ",", fixed = TRUE)))
})
```

This function is similar to the code from the previous 2 lines. The only difference is in the `split` parameter which is `|` above and `,` here.

Reply:
I will account for this.